circuits_over_bits.md

```diff
@@ -0,0 +1,199 @@
+q_1: |0>┤ H ├
+        ├───┤
+q_2: |0>┤ H ├
+        └───┘
+q_3: |0>─────
+             
+q_4: |0>─────
+             
+q_5: |0>─────
+             
+```
+
+Note that in the 2nd case, a register is still defined under the hood (register `q`). This was done for backwards
+compatibility, since much of Qiskit's internal code uses `circuit.qregs`.
+
+This RFC is for fully migrating to circuits being defined over `Qubit`s and `Clbit`s, and making `QuantumRegister` and
+`ClassicalRegister` defined as pointers to circuit bits. The goal is to simplify internal usage of circuits
+by making Qubits/Clbits into fundamental building blocks of circuits, and only defining registers when the need arises.
+This would also help circuits maintain a canonical ordering of their bits, which is more natural for the datastructures
+that define circuits.
+
+## Requirement 1: Still support circuits over registers
+
+Some circuits naturally benefit from a definition of registers. Examples include:
+1. Error correction circuits, which are defined over ancilla and data registers
+2. Multi-controlled circuits, which might have a control register and a target register
+3. Arithmetic circuits, like a full adder, which is defined over to registers representing an integer each.
+
+## Requirement 2: Circuits aware of registers if they exist
+
+One design option is to make circuits only aware of their `Qubit`s/`Clbit`s, and make registers 3rd party pointers into
+those. We explicitly rule out this option, since we sometimes want the circuit to be aware of registers too, if they
+exist. Usage examples include:
+1. Drawing a circuit, and visualizing its registers
+2. Exposing a library of circuits, with registers exposed as well for ease of use
+
+## Design
+We propose some changes to Terra that will facilitate the following programs, annotated for clarity:
+
+```python
+# you can assign circuit qubits to registers upon creation of the circuit.
+# These registers are now nothing but lists containing the circuit's qubits
+qr_a = QuantumRegister(3, 'a')
+qr_b = QuantumRegister(3, 'b')
+circuit = QuantumCircuit(qr_a, qr_b)
+
+# this can alternatively (and more explicitly) be written as
+qr_a = QuantumRegister(name='a', bits=3*[Qubit])
+qr_b = QuantumRegister(name='b', bits=3*[Qubit])
+circuit = QuantumCircuit(qr_a, qr_b)
+
+# the same outcome can be achieved by assigning registers after circuit creation
+circuit = QuantumCircuit(6)   # a register-less circuit over 6 qubits
+print(circuit.qubits)
+>> [Qubit(), Qubit(), Qubit(), Qubit(), Qubit(), Qubit()]
+print(circuit.qregs)
+>> {}
+
+q = circuit.qubits
+circuit.mark_register([q[0], q[1], q[2]], 'a')  # now tag the bits as an 'a' and 'b' register
+circuit.mark_register([q[2], q[3], q[4]], 'b')  # (this info will be kept in circuit)
+print(circuit.qregs)
+>> {'a': QuantumRegister(3, 'a'),
+    'b': QuantumRegister(3, 'b')}
+
+circuit.mark_register([q[0]], 'c')  # marking a bit as belonging to more than one register is disallowed
```

Comment:
In the example above `q[2]` is marked as belonging to both registers `'a'` and `'b'`, was it meant to be `circuit.mark_register([q[3], q[4], q[5]], 'b')`?

Reply:
thanks, fixed